packages/apputils/src/dialog.ts

```diff
@@ -34,26 +34,23 @@ import {
  * @returns A promise that resolves with whether the dialog was accepted.
  */
 export
-function showDialog(options: Dialog.IOptions={}): Promise<Dialog.IButton> {
+function showDialog<T>(options: Dialog.IOptions<T>={}): Promise<Dialog.IResult<T>> {
   let dialog = new Dialog(options);
-  return dialog.launch().then(result => {
-    dialog.dispose();
-    return result;
-  });
+  return dialog.launch();
 }
 
 
 /**
  * A modal dialog widget.
  */
 export
-class Dialog extends Widget {
+class Dialog<T> extends Widget {
   /**
    * Create a dialog panel instance.
    *
    * @param options - The dialog setup options.
    */
-  constructor(options: Dialog.IOptions={}) {
+  constructor(options: Dialog.IOptions<T>={}) {
     super();
     this.addClass('jp-Dialog');
     options = Private.handleOptions(options);
```

Comment:
This form of option handling is not `null` safe. The type of `options` is still an object with all optional properties.